addonscript/src/addonscript/api.py

```diff
@@ -45,29 +52,65 @@ async def add_version(context, version):
     return result
 
 
+async def check_upload(context, uuid):
+    """Check on the status of file upload identified by `uuid`
+
+    Raises:
+        SignatureError: If the upload is awaiting processing
+        FatalSignatureError: If validation errors occurred
+    """
+    url = get_api_url(context, UPLOAD_DETAIL, uuid=uuid)
+    result = await amo_get(context, url)
+    if not result["processed"]:
+        # retry
+        raise SignatureError("upload not yet processed")
+    if not result["valid"]:
+        raise FatalSignatureError(f"Automated validation produced errors: {result['validation']}")
+
+
 async def do_upload(context, locale):
     """Upload the language pack for `locale` to AMO.
 
     Returns the JSON response from AMO
     """
     locale_info = context.locales[locale]
-    langpack_id = locale_info["id"]
-    version = locale_info["version"]
-    url = get_api_url(context, UPLOAD_VERSION, id=langpack_id, version=version)
-    with open(context.locales[locale]["unsigned"], "rb") as file:
+    url = get_api_url(context, UPLOAD_CREATE)
+    with open(locale_info["unsigned"], "rb") as file:
         data = {"channel": get_channel(context.task), "upload": file}
-        try:
-            result = await amo_put(context, url, data)
-        except ClientResponseError as exc:
-            # XXX: .code is deprecated in aiohttp 3.1 in favor of .status
-            if exc.status == 409:
-                raise AMOConflictError("Addon <{}> already present on AMO with version <{}>".format(langpack_id, version))
-            # If response code is not 409 - CONFLICT, bubble the exception
-            raise exc
-        return result
-
-
-async def get_signed_addon_url(context, locale, pk):
+        return await amo_post(context, url, data)
+
+
+async def do_create_version(context, locale, upload_uuid):
+    """Create a new version for the `locale` language pack on AMO.
+
+    Returns the version's identifier
+
+    Raises:
+        AMOConflictError: If the version already exists for this addon
+    """
+    locale_info = context.locales[locale]
+    langpack_id = locale_info["id"]
+    url = get_api_url(context, ADDON_CREATE_OR_EDIT, id=langpack_id)
+    data = {
+        "categories": {"firefox": ["language"]},
+        "name": locale_info["name"],
+        "summary": locale_info["description"],
+        "version": {
+            "license": "MPL-2.0",
+            "upload": upload_uuid,
+        },
```

Comment:
I'm hardcoding the application name and license here, because that's not explicit in the xpi.
I think this is fine for right now because there's a few other places we already assume firefox, and it seems unlikely we'd need to change the license for specific langpacks (and if it happens we can reconsider then and add that as part of the task payload or something).

Reply:
It might be worthwhile to not assume it's Firefox. Now that we can use the newer API, maybe we can automate upload for the web-extensions in the add-ons pipeline? IIRC there's a few add-ons in mozilla-extensions that are uploaded to AMO manually. This is fine for now, add-on script is only used by langpacks in Gecko.